### ISSUE TITLE
Improper intrinsic track sizing for multi-line Text

### ISSUE THREAD

**Issue by klondikedragon** [shyndman/flutter_layout_grid#76]
This widget is extremely powerful and very handy, thank you!

Imagine you have a vertical `CustomScrollView` with a `SliverFillRemaining` sliver (`hasScrollBody`=`false`) that contains a container with a minimum height set to 300px that contains a `LayoutGrid` with a single column and 2 rows:
* A flex row that expands to whatever free space is remaining
* A row that contains a Text that is long enough to wrap across multiple lines

Yes, this is a complicated layout! Essentially, it's a simplified example of an application that has a lot of vertically expanding content that is intrinsically sized, and then needs one row in a grid that is normally sized to fit the rest of the viewport, but if that content is longer than the height of the viewport it will then be able to scroll to see everything.

The bug in this situation is that in the first pass Flutter will call `computeMaxIntrinsicHeight` on the `RenderLayoutGrid` renderbox (because it's in a context where the cross-axis (horizontal) is fixed and it wants to know how high the widget can go for a given width), and flutter passes the fixed cross-axis width. The code is currently specifying a constraint of a minimum width for the given width passed to `_computeIntrinsicSize`: 

```
  @override
  double computeMaxIntrinsicHeight(double width) =>
      _computeIntrinsicSize(BoxConstraints(minWidth: width)).maxTracksHeight;
```

`_computeIntrinsicSize` then calls computeGridSize passing the constraints, but leaving the child constraints set to null. Then the child constraints land up getting inflated to infinity.

This is a problem because when it's computing the row track height, even with a grid fit of passthrough, the childConstraints has a min width of the defined amount but a max width of infinity. The max width of infinity gets translated into a `crossAxisSize=0` when `minIntrinsicSize` is finally called. And because `crossAxisSize=0`, then the Text widget computes a height of like 3000 pixels (exact number depends on how much text is passed) because the Text RenderBox is asked what it's intrinsic height would be with a zero width, and because Text will wrap it creates one line of text per word making it very big. This leads to a very large max intrinsic height of thousands of pixels. Then flutter calls the layout itself, and now it passes the constraint with the height set to 3000+ pixels. The width is now set in this second computation, and the computeGridSize call this time only needs a few lines of text (for example). This leaves the flex row expanding to be thousands of pixels in size unnecessarily.

Here is the debug track layout print output with some extra stuff added showing the problem (in this case there is a single column with 3 rows, row 0 and row 2 are intrinsically sized, and row 2 has the multi-line text widget):

<details>
 <summary>Debug Log Showing Problem</summary>

```
DEBUG: ============= computeMaxIntrinsicHeight(579) =========
DEBUG: ------------------------- computeGridSize(gridFit=GridFit.passthrough, gridConstraints=BoxConstraints(579.0<=w<=Infinity, 0.0<=h<=Infinity), childConstraints=null)
DEBUG: ------------------------- computeGridSize updated childConstraints=BoxConstraints(579.0<=w<=Infinity, 0.0<=h<=Infinity)
DEBUG COLUMN tracks with a maximum free space of 0, isAxisUpperBound=false
min-max: 0.0->0.0 (same)
free space: 0
DEBUG ROW tracks with a maximum free space of 0, isAxisUpperBound=false
Resolving intrinsic row heights [tracks [0,2]] with constraints BoxConstraints(579.0<=w<=Infinity, 0.0<=h<=Infinity)
DEBUG: measuring minIntrinsicSize for 1 items...
DEBUG: measuring min intrinsic item size on axis Axis.vertical with crossAxisSize=0
  min size of track 0 = 56
  distributing 56 across track 0 on min
  max size of track 0 = 56
DEBUG: measuring minIntrinsicSize for 1 items...
DEBUG: measuring min intrinsic item size on axis Axis.vertical with crossAxisSize=0
  min size of track 2 = 3902
  distributing 3902 across track 2 on min
  max size of track 2 = 3902
  update track 0 = 56.0
  update track 2 = 3902.0
min-max: 3958.0->3958.0 (same)
free space: -3958
DEBUG ================= Starting grid layout for constraints BoxConstraints(w=579.0, h=3958.0), child constraints BoxConstraints(w=579.0, h=3958.0)
DEBUG: ------------------------- computeGridSize(gridFit=GridFit.passthrough, gridConstraints=BoxConstraints(w=579.0, h=3958.0), childConstraints=null)
DEBUG: ------------------------- computeGridSize updated childConstraints=BoxConstraints(w=579.0, h=3958.0)
DEBUG COLUMN tracks with a maximum free space of 579, isAxisUpperBound=true
min-max: 0.0->0.0 (same)
free space: 579
DEBUG ROW tracks with a maximum free space of 3958, isAxisUpperBound=true
Resolving intrinsic row heights [tracks [0,2]] with constraints BoxConstraints(w=579.0, h=3958.0)
DEBUG: measuring minIntrinsicSize for 1 items...
DEBUG: measuring min intrinsic item size on axis Axis.vertical with crossAxisSize=579
  min size of track 0 = 56
  distributing 56 across track 0 on min
  max size of track 0 = 56
DEBUG: measuring minIntrinsicSize for 1 items...
DEBUG: measuring min intrinsic item size on axis Axis.vertical with crossAxisSize=579
  min size of track 2 = 94
  distributing 94 across track 2 on min
  max size of track 2 = 94
  update track 0 = 56.0
  update track 2 = 94.0
min-max: 150.0->150.0 (same)
free space: 3808
Determined track sizes:
  column 0: 579
  row 0: 56
  row 1: 3808
  row 2: 94
Finished track sizing
```
</details>

The solution is pretty simple, when computing the max intrinsic width or max intrinsic height, instead of setting a `minHeight` or `minWidth` constraint, set a `maxHeight` or `maxWidth` constraint:

```
  @override
  double computeMaxIntrinsicWidth(double height) =>
      _computeIntrinsicSize(BoxConstraints(maxHeight: height)).maxTracksWidth;

  @override
  double computeMaxIntrinsicHeight(double width) =>
      _computeIntrinsicSize(BoxConstraints(maxWidth: width)).maxTracksHeight;
```

This logic follows by reading the doc comments where it talks about width-in-height-out layout algorithms (e.g., which is what the `Text` widget is with line wrapping).

With this fix in place, the layout works exactly as desired, with no extra space in the flex widget. Here is the debug layout log with the fix in place:

<details>
 <summary>Debug Log With Fix In Place</summary>

```
DEBUG: ============= computeMaxIntrinsicHeight(579) =========
DEBUG: ------------------------- computeGridSize(gridFit=GridFit.passthrough, gridConstraints=BoxConstraints(0.0<=w<=579.0, 0.0<=h<=Infinity), childConstraints=null)
DEBUG: ------------------------- computeGridSize updated childConstraints=BoxConstraints(0.0<=w<=579.0, 0.0<=h<=Infinity)
DEBUG COLUMN tracks with a maximum free space of 579, isAxisUpperBound=true
min-max: 0.0->0.0 (same)
free space: 579
DEBUG ROW tracks with a maximum free space of 0, isAxisUpperBound=false
Resolving intrinsic row heights [tracks [0,2]] with constraints BoxConstraints(0.0<=w<=579.0, 0.0<=h<=Infinity)
DEBUG: measuring minIntrinsicSize for 1 items...
DEBUG: measuring min intrinsic item size on axis Axis.vertical with crossAxisSize=579
  min size of track 0 = 56
  distributing 56 across track 0 on min
  max size of track 0 = 56
DEBUG: measuring minIntrinsicSize for 1 items...
DEBUG: measuring min intrinsic item size on axis Axis.vertical with crossAxisSize=579
  min size of track 2 = 94
  distributing 94 across track 2 on min
  max size of track 2 = 94
  update track 0 = 56.0
  update track 2 = 94.0
min-max: 150.0->150.0 (same)
free space: -150
DEBUG ================= Starting grid layout for constraints BoxConstraints(w=579.0, h=469.0), child constraints BoxConstraints(w=579.0, h=469.0)
DEBUG: ------------------------- computeGridSize(gridFit=GridFit.passthrough, gridConstraints=BoxConstraints(w=579.0, h=469.0), childConstraints=null)
DEBUG: ------------------------- computeGridSize updated childConstraints=BoxConstraints(w=579.0, h=469.0)
DEBUG COLUMN tracks with a maximum free space of 579, isAxisUpperBound=true
min-max: 0.0->0.0 (same)
free space: 579
DEBUG ROW tracks with a maximum free space of 469, isAxisUpperBound=true
Resolving intrinsic row heights [tracks [0,2]] with constraints BoxConstraints(w=579.0, h=469.0)
DEBUG: measuring minIntrinsicSize for 1 items...
DEBUG: measuring min intrinsic item size on axis Axis.vertical with crossAxisSize=579
  min size of track 0 = 56
  distributing 56 across track 0 on min
  max size of track 0 = 56
DEBUG: measuring minIntrinsicSize for 1 items...
DEBUG: measuring min intrinsic item size on axis Axis.vertical with crossAxisSize=579
  min size of track 2 = 94
  distributing 94 across track 2 on min
  max size of track 2 = 94
  update track 0 = 56.0
  update track 2 = 94.0
min-max: 150.0->150.0 (same)
free space: 319
Determined track sizes:
  column 0: 579
  row 0: 56
  row 1: 319
  row 2: 94
Finished track sizing
```
</details>

I'm pretty sure the changes to `computeMaxIntrinsicWidth` and `computeMaxIntrinsicHeight` are correct in all cases, but you could also choose to only use the new behavior based on a new flag to the `LayoutGrid`.

**Comment by klondikedragon** [comment]
Upon further experimentation, these changes do change the visual appearance of the Periodic Table example. I'll submit a PR for review shortly that adds flags to the widget that will allow activation of a "width-in-height-out" or "height-in-width-out" layout approach (both defaulting to false, to use the previous behavior).

**Comment by shyndman** [comment]
Wow! Thanks for being so detailed. I'm just going to need a couple more days to give this the attention it deserves.

What is it that you're building, if you're willing to share?

**Comment by klondikedragon** [comment]
Thanks @shyndman -- the changes in PR #77 have been working out nicely so far to fix the issue. Hopefully should be enough to fix the issue without introducing any new ones.

I've found this grid is really handy for building responsive layouts. When the viewport is smaller than a given width, I have it lay out three rows in one column:

![image](https://user-images.githubusercontent.com/88254524/162859481-0444dbd9-201f-41ad-ae8b-9b8972b444d4.png)

The first and third rows are intrinsically sized, and the second row is a flex row. The third row is the one whose intrinsic height changes with the content. For example, if the detail text is longer, it wraps and takes up more vertical space. And if the height of the viewport is not large enough, it will have a vertical scroll bar. In this case, the fluent_ui treeview list has a minimum height and will start to scroll if it can't contain all items, and then if the 3 rows themselves all can't fit within the viewport, the viewport itself starts to show a vertical scrollbar (note the two vertical scrollbars in this example):

![image](https://user-images.githubusercontent.com/88254524/162859718-0c6195c3-8a35-4de1-ada1-b22e7a47a29b.png)

Alternatively, if the viewport is wide enough, then it switches to a mode with two rows -- the first row spans two columns, and then the second row has a different widget in each column (first column is fixed-width, the second column flexes to fit to size of viewport)

![image](https://user-images.githubusercontent.com/88254524/162860013-e268d95b-2c2c-4ba9-81d3-18f542946d3e.png)

I like the approach of this package vs some others that implement the "bootstrap" layout model. Using ASCII art to show where you want rows/columns to go is really intuitive, and there is more precise control over width/height (with bootstrap all elements will "expand" proportionally as window expands, but in this case the left pane is more of a side bar and better to stay (mostly) fixed width while the right pane expands to fit (for example).

**Comment by shyndman** [comment]
Looks great! I use the grid for exactly the same purpose.

I went over the issue, and I agree with your fix. I left one nit on the PR — add the comment and I'll merge.

Now I'm wondering if this is a breaking change or not. What do you think?

**Comment by klondikedragon** [comment]
It seems like it *could* be breaking, but IMO only in a good way. The behavior before wouldn't really be desirable if you were trying to use in conjunction with a widget that determined it's height dynamically based on a fixed width (like `Wrap` or `Text`). On the other hand, no one has complained before. So it could be safer to bump the major version in pub.dev package just in case?

**Comment by shyndman** [comment]
Yeah, agreed. Safer. Give me a few and I'll get this published.

**Comment by shyndman** [comment]
Published as v2.0.0.

Thanks for the help.

**Comment by klondikedragon** [comment]
@shyndman - Perfect, thank you!
